source/common/quic/envoy_quic_client_stream.cc

```diff
@@ -80,30 +80,38 @@ void EnvoyQuicClientStream::encodeData(Buffer::Instance& data, bool end_stream)
   ASSERT(!local_end_stream_);
   local_end_stream_ = end_stream;
   SendBufferMonitor::ScopedWatermarkBufferUpdater updater(this, this);
-  Buffer::RawSliceVector raw_slices = data.getRawSlices();
-  absl::InlinedVector<quiche::QuicheMemSlice, 4> quic_slices;
-  quic_slices.reserve(raw_slices.size());
-  for (auto& slice : raw_slices) {
-    ASSERT(slice.len_ != 0);
-    // Move each slice into a stand-alone buffer.
-    // TODO(danzh): investigate the cost of allocating one buffer per slice.
-    // If it turns out to be expensive, add a new function to free data in the middle in buffer
-    // interface and re-design QuicheMemSliceImpl.
-    quic_slices.emplace_back(quiche::QuicheMemSlice::InPlace(), data, slice.len_);
-  }
-  quic::QuicConsumedData result{0, false};
-  absl::Span<quiche::QuicheMemSlice> span(quic_slices);
-  {
+  if (http_datagram_handler_) {
     IncrementalBytesSentTracker tracker(*this, *mutableBytesMeter(), false);
-    result = WriteBodySlices(span, end_stream);
-  }
-  // QUIC stream must take all.
-  if (result.bytes_consumed == 0 && has_data) {
-    IS_ENVOY_BUG(fmt::format("Send buffer didn't take all the data. Stream is write {} with {} "
-                             "bytes in send buffer. Current write was rejected.",
-                             write_side_closed() ? "closed" : "open", BufferedDataBytes()));
-    Reset(quic::QUIC_BAD_APPLICATION_PAYLOAD);
-    return;
+    if (!http_datagram_handler_->encodeCapsuleFragment(data.toString(), end_stream)) {
+      Reset(quic::QUIC_BAD_APPLICATION_PAYLOAD);
+      return;
+    }
+  } else {
```

Comment:
I think we should have @RyanTheOptimist or @danzh2010 review this as I'm not likely to catch bugs

Reply:
Thanks for tagging Ryan and Dan. I will appreciate any comments from you all.